internal/controllers/configuration/konglicense_controller.go

```diff
@@ -0,0 +1,347 @@
+package configuration
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/google/uuid"
+	"github.com/kong/go-kong/kong"
+	"github.com/samber/lo"
+	"github.com/samber/mo"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	k8stypes "k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/cache"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/kong/kubernetes-ingress-controller/v3/internal/controllers"
+	"github.com/kong/kubernetes-ingress-controller/v3/internal/util"
+	"github.com/kong/kubernetes-ingress-controller/v3/internal/util/kubernetes/object/status"
+	kongv1alpha1 "github.com/kong/kubernetes-ingress-controller/v3/pkg/apis/configuration/v1alpha1"
+)
+
+// -----------------------------------------------------------------------------
+// KongV1Alpha1 KongLicense - Reconciler
+// -----------------------------------------------------------------------------
+
+// KongV1Alpha1KongLicenseReconciler reconciles KongLicense resources.
+type KongV1Alpha1KongLicenseReconciler struct {
+	client.Client
+
+	Log              logr.Logger
+	Scheme           *runtime.Scheme
+	LicenseCache     cache.Store
+	CacheSyncTimeout time.Duration
+	StatusQueue      *status.Queue
+	// ControllerName is the name of the controller to use in `controllerName` field in controller status item.
+	ControllerName string
+
+	chosenLicenseLock sync.RWMutex
+	chosenLicense     *kongv1alpha1.KongLicense
+}
+
+const (
+	// LicenseControllerType annotates the controller type.
+	LicenseControllerType = "konghq.com/kong-ingress-controller"
+)
+
+const (
+	ConditionTypeProgrammed = "Programmed"
+	// ConditionReasonPickedAsLatest represents that the KongLicense being picked as the newest one.
+	ConditionReasonPickedAsLatest = "PickedAsLatest"
+	// ConditionReasonReplacedByNewer represents that the KongLicense is replaced by other one that is newer.
+	ConditionReasonReplacedByNewer = "ReplacedByNewer"
+	// maxConditionNum is the maximum number of condition items in the controller status.
+	maxConditionNum = 8
+)
+
+var _ controllers.Reconciler = &KongV1Alpha1KongLicenseReconciler{}
+
+func NewLicenseCache() cache.Store {
+	return cache.NewStore(kongLicenseKeyFunc)
+}
```

Comment:
Because the licenses are not processed inside the translator but used as a license injector, so I do not want to add them into the general "big cache".

Reply:
Ah, I see, the reconciler implements `GetLicense` and it's going to be used in the `Translator`. Let's make sure that we draw the boundary between the reconciler and the translator well by defining an interface. 👍